File: Data Structures & Algorithms/permutations/submission-0.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        def path(temp):
            if len(temp) == len(nums):
                res.append(temp[:])

            for num in nums:
                if num in temp:
                    continue
                temp.append(num)
                path(temp)
                temp.pop()

        path([])
        return res
        
        

File: Data Structures & Algorithms/three-integer-sum/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    threeSum(nums) {
        nums.sort((a,b)=> a-b);
        const notebook = {};
        nums.forEach((num,index)=>{
            if(!notebook[-num]){
                notebook[-num] = [];
            }
            notebook[-num].push(index);
        });

        const solutionSet = new Set();
        const res = [];

        for(let i = 0; i< nums.length;i++){
            for(let j = 2; j<nums.length;j++){
                const sum = nums[i] + nums[j];
                if(notebook[sum]){
                    notebook[sum].forEach((index)=>{
                        if(i<index && index < j){
                            const solutionKey = `${nums[i]}${nums[index]}${nums[j]}}`;
                            if(!solutionSet.has(solutionKey)){
                                solutionSet.add(solutionKey);
                                res.push([nums[i],nums[index],nums[j]]);
                            }
                        }
                    })
                }
            }
        }
        return res;
    }
}
